man/tex-fmt.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH tex-fmt 1  "tex-fmt 0.5.7" 
.SH NAME
tex\-fmt \- LaTeX formatter written in Rust
.SH SYNOPSIS
\fBtex\-fmt\fR [\fB\-c\fR|\fB\-\-check\fR] [\fB\-p\fR|\fB\-\-print\fR] [\fB\-f\fR|\fB\-\-fail\-on\-change\fR] [\fB\-n\fR|\fB\-\-nowrap\fR] [\fB\-l\fR|\fB\-\-wraplen\fR] [\fB\-t\fR|\fB\-\-tabsize\fR] [\fB\-\-usetabs\fR] [\fB\-s\fR|\fB\-\-stdin\fR] [\fB\-\-config\fR] [\fB\-\-noconfig\fR] [\fB\-v\fR|\fB\-\-verbose\fR] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-\-trace\fR] [\fB\-\-completion\fR] [\fB\-\-man\fR] [\fB\-\-args\fR] [\fB\-r\fR|\fB\-\-recursive\fR] [\fB\-\-format\-tables\fR] [\fB\-\-wrap\-by\-visual\-len\fR] [\fB\-\-wrap\-cjk\fR] [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fIfiles\fR] 
.SH DESCRIPTION
LaTeX formatter written in Rust
.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-check\fR
Check formatting, do not modify files
.TP
\fB\-p\fR, \fB\-\-print\fR
Print to stdout, do not modify files
.TP
\fB\-f\fR, \fB\-\-fail\-on\-change\fR
Format files and return non\-zero exit code if files are modified
.TP
\fB\-n\fR, \fB\-\-nowrap\fR
Do not wrap long lines
.TP
\fB\-l\fR, \fB\-\-wraplen\fR \fI<N>\fR
Line length for wrapping [default: 80]
.TP
\fB\-t\fR, \fB\-\-tabsize\fR \fI<N>\fR
Number of characters to use as tab size [default: 2]
.TP
\fB\-\-usetabs\fR
Use tabs instead of spaces for indentation
.TP
\fB\-s\fR, \fB\-\-stdin\fR
Process stdin as a single file, output to stdout
.TP
\fB\-\-config\fR \fI<PATH>\fR
Path to config file
.TP
\fB\-\-noconfig\fR
Do not read any config file
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Show info messages
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Hide warning messages
.TP
\fB\-\-trace\fR
Show trace messages
.TP
\fB\-\-completion\fR \fI<SHELL>\fR
Generate shell completion script
.br

.br
\fIPossible values:\fR
.RS 14
.IP \(bu 2
bash
.IP \(bu 2
elvish
.IP \(bu 2
fish
.IP \(bu 2
powershell
.IP \(bu 2
zsh
.RE
.TP
\fB\-\-man\fR
Generate man page
.TP
\fB\-\-args\fR
Print arguments passed to tex\-fmt and exit
.TP
\fB\-r\fR, \fB\-\-recursive\fR
Recursively search for files to format
.TP
\fB\-\-format\-tables\fR
Format tables
.TP
\fB\-\-wrap\-by\-visual\-len\fR
Use visual length instead of character count when wrapping
.TP
\fB\-\-wrap\-cjk\fR
Allow wrap at CJK characters
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.TP
[\fIfiles\fR]
List of files to be formatted
.SH VERSION
v0.5.7
.SH AUTHORS
William George Underwood, wg.underwood13@gmail.com


NEWS.md
# v0.5.7

- New `--format-tables` option to align ampersands and insert line breaks in table environments.
- Now `" "` must be explicitly passed to `wrap-chars` to enable line wrapping at spaces.
- Allow for `\verb+...+` syntax.
- Use `usize` for `wraplen` and `wrapmin` to permit wrapping at longer line lengths.
- Upload `latex-formatter` package to CTAN.
- Ensure `wasm` feature is built for the web app.
- Fix bug in recursive search when file extensions are missing.

# v0.5.6

- Gate web functionality behind a crate feature.
- Fix bug with `wrapmin` configuration option being ignored.
- Handle `\mintinline` as `\verb`.
- Fix bug with file extension handling

# v0.5.5

- Implement recursive file formatting with `--recursive` or `-r`.
- Format current directory with `tex-fmt` (no explicit files passed).
- Fix bug when wrapping lines containing non-ASCII characters.
- Add warning for first negative indent, if any.
- Improve performance by using `RegexSet` for multiple matches.
- Faster indent calculation by avoiding repeated delimiter counting.
- File names handled internally with `PathBuf` rather than `String`.

# v0.5.4

## Allow custom non-indented environments

- Pass `no-indent-envs = ["mydocument"]` in `tex-fmt.toml`.
- Then no indentation is applied inside `\begin{mydocument}...\end{mydocument}`.
- This is the default behaviour for `\begin{document}...\end{document}`.

## Allow custom verbatim environments

- Pass `verbatims = ["myverbatim"]` in `tex-fmt.toml`.
- All formatting is then skipped inside `\begin{myverbatim}...\end{myverbatim}`.

## Improve formatting when using `\verb|...|`

- Lines are not broken inside `\verb|...|`.
- Environments inside `\verb|...|` do not trigger new lines.
- No indenting applied to lines containing `\verb|...|`.

## Minor changes

- Fix output bug in profiling script at `extra/prof.sh`.
- Add link to [docs.rs](https://docs.rs/tex-fmt/latest/tex_fmt/) page in README.
- Add logo to [docs.rs](https://docs.rs/tex-fmt/latest/tex_fmt/) page.
- Add better tests for config file options and CLI arguments.
- Improve documentation of config file options and CLI arguments in README.

# v0.5.3

- Add `--fail-on-change` flag.
- Deploy web app for using tex-fmt in a browser.
- Simplify testing structure.
- Switch nix flake input to nixpkgs-unstable.
- Update README with GitHub Action, mason.nvim, Debian, bibtex-tidy, pre-commit.

# v0.5.2

- Fix critical bug with config files missing the `lists` field.
- Trim trailing newlines.

# v0.5.1

- Custom list environments can be passed using the `lists` option in the config file.
- Allow `verbosity = "info"` in the config file.
- Fixed a bug with configuration values being incorrectly reset.

# v0.5.0

Version v0.5.0 is a major release, including breaking changes and substantial new features.

## Changes to existing CLI options
- The option to disable line wrapping has been changed from `--keep` to `--nowrap`.
- The option to set the number of characters used per indentation level has been changed from `--tab` to `--tabsize`.
- The option to set the maximum line length for wrapping has been changed from `--wrap` to `--wraplen`.
- See below for information on the new `--config`, `--noconfig`, `--man`, `--completion`, and `--args` flags.

## Configuration file support
Configuring tex-fmt can now be achieved using a configuration file as well as CLI arguments. The configuration file can be read from a user-specified path with `--config <config>`, from the current working directory, from the root of the current git repository, or from the user's configuration directory, in order of decreasing priority. Arguments passed on the command line will always override those specified in configuration files. Configuration files can be disabled by passing `--noconfig`.

## Man pages
Man pages can be generated using the `--man` flag. Pre-built man pages are also available for download from the GitHub repository.

## Shell completion
Completion files for popular shells, including bash, fish, zsh, elvish and PowerShell, can be generated using the `--completion <SHELL>` flag. Pre-built completion scripts are also available for download from the GitHub repository.

## Minor changes
- Arguments passed to tex-fmt can be inspected by passing `--args`
- Fixed bug with `\itemsep` matching the `\item` pattern
- Added last non-indented line number to "Indent did not return to zero" error messages
- Removed LTO optimization to improve compile time with minimal effect on run time
- If duplicate file names are provided, they are now removed before formatting
- Added LLF to the list of existing tools
- Changed order of options in help dialogs

# v0.4.7

- Fix bug with `--stdin` adding newlines at EOF
- Fix logic for ignoring verbatim environments
- Ensure sectioning commands begin on new lines
- Various performance improvements
- Add NEWS.md for release notes
- Ensure all test files successfully compile to PDFs
- Better documentation of options in README.md

# v0.4.6

- Added `--wrap` flag to choose line length for wrapping
- Significant changes to central formatting logic to reduce memory allocations
- Treat comment environments as verbatim
- Improved performance with finding comments in source code

# v0.4.5

- Added `--usetabs` to use tabs instead of spaces for indentation
- Fixed a bug with unicode graphemes and comment handling
- Main function now returns `std::process::ExitCode` for a cleaner exit
- Reduced memory allocation in comment handling logic
- Reduced memory allocation when indenting lines
- Caching of pattern matches reduces number of regex searches

# v0.4.4

- Added `--tab` flag for variable tab size [default: 2]
- Fixed bug with incorrect line numbers being printed
- Fixed bug with quadratic complexity of comment checking
- Added Arch User Repository support
- Added VS Code support
- Improved performance by moving environment checking inside main loop
- Upgraded Cargo dependencies
- Included LTO optimization on the release build


completion/tex-fmt.bash
_tex-fmt() {
    local i cur prev opts cmd
    COMPREPLY=()
    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
        cur="$2"
    else
        cur="${COMP_WORDS[COMP_CWORD]}"
    fi
    prev="$3"
    cmd=""
    opts=""

    for i in "${COMP_WORDS[@]:0:COMP_CWORD}"
    do
        case "${cmd},${i}" in
            ",$1")
                cmd="tex__fmt"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        tex__fmt)
            opts="-c -p -f -n -l -t -s -v -q -r -h -V --check --print --fail-on-change --nowrap --wraplen --tabsize --usetabs --stdin --config --noconfig --verbose --quiet --trace --completion --man --args --recursive --format-tables --wrap-by-visual-len --wrap-cjk --help --version [files]..."
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --wraplen)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -l)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --tabsize)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -t)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --config)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --completion)
                    COMPREPLY=($(compgen -W "bash elvish fish powershell zsh" -- "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then
    complete -F _tex-fmt -o nosort -o bashdefault -o default tex-fmt
else
    complete -F _tex-fmt -o bashdefault -o default tex-fmt
fi
